amazon/ion/ioncmodule.c

```diff
@@ -0,0 +1,1369 @@
+    PyObject* ion_type = NULL;
+    if (PyObject_HasAttrString(obj, "ion_type")) {
+        ion_type = PyObject_GetAttrString(obj, "ion_type");
+    }
+    if (ion_type == NULL) return tid_none_INT;
+    int c_type = c_ion_type_table[PyInt_AsSsize_t(ion_type)];
+    Py_DECREF(ion_type);
+    return c_type;
+}
+
+/*
+ *  Gets a C string from a python string
+ *
+ *  Args:
+ *      str:  A python string that needs to be converted
+ *      out:  A C string converted from 'str'
+ *      len_out:  Length of 'out'
+ */
+static iERR c_string_from_py(PyObject* str, char** out, Py_ssize_t* len_out) {
+    iENTER;
+#if PY_MAJOR_VERSION >= 3
+    *out = PyUnicode_AsUTF8AndSize(str, len_out);
+#else
+    PyObject *utf8_str;
+    if (PyUnicode_Check(str)) {
+        utf8_str = PyUnicode_AsUTF8String(str);
+    }
+    else {
+        utf8_str = PyString_AsEncodedObject(str, "utf-8", "strict");
+    }
+    if (!utf8_str) {
+        _FAILWITHMSG(IERR_INVALID_ARG, "Python 2 fails to convert python string to utf8 string.");
+    }
+    PyString_AsStringAndSize(utf8_str, out, len_out);
+    Py_DECREF(utf8_str);
+#endif
+    iRETURN;
+}
+
+/*
+ *  Gets an ION_STRING from a python string
+ *
+ *  Args:
+ *      str:  A python string that needs to be converted
+ *      out:  An ION_STRING converted from 'str'
+ */
+static iERR ion_string_from_py(PyObject* str, ION_STRING* out) {
+    iENTER;
+    char* c_str = NULL;
+    Py_ssize_t c_str_len;
+    IONCHECK(c_string_from_py(str, &c_str, &c_str_len));
+    ION_STRING_INIT(out);
+    ion_string_assign_cstr(out, c_str, c_str_len);
+    iRETURN;
+}
+
+/*
+ *  Builds a python string using an ION_STRING
+ *
+ *  Args:
+ *      string_value:  An ION_STRING that needs to be converted
+ *
+ *  Returns:
+ *      A python string
+ */
+static PyObject* ion_build_py_string(ION_STRING* string_value) {
+    // TODO Test non-ASCII compatibility.
+    // NOTE: this does a copy, which is good.
+    if (!string_value->value) return Py_None;
+    return PyUnicode_FromStringAndSize((char*)(string_value->value), string_value->length);
+}
+
+/*
+ *  Adds an element to a List or struct
+ *
+ *  Args:
+ *      pyContainer:  A container that the element is added to
+ *      element:  The element to be added to the container
+ *      in_struct:  if the current state is in a struct
+ *      field_name:  The field name of the element if it is inside a struct
+ */
+static void ionc_add_to_container(PyObject* pyContainer, PyObject* element, BOOL in_struct, ION_STRING* field_name) {
+    if (in_struct) {
+        PyObject_CallMethodObjArgs(
+            pyContainer,
+            PyString_FromString("add_item"),
+            ion_build_py_string(field_name),
+            (PyObject*)element,
+            NULL
+        );
+    }
+    else {
+        PyList_Append(pyContainer, (PyObject*)element);
+    }
+    Py_XDECREF(element);
+}
+
+/*
+ *  Converts an ion decimal string to a python-decimal-accept string. NOTE: ion spec uses 'd' in a decimal number
```

Comment:
Python decimal accepts string with char 'e'. e.g. `"123e456"`.

`ion_decimal_to_string` API returns string with char 'd'. e.g. `"123d456"` which fails python decimal constructor.

This can be solve after we calculate decimal without using strings.